src/main/kotlin/uk/gov/justice/digital/hmpps/managerecallsapi/service/PartBRecordService.kt

```diff
@@ -0,0 +1,112 @@
+) {
+
+  @Transactional
+  fun createRecord(
+    recallId: RecallId,
+    currentUserId: UserId,
+    request: PartBRequest
+  ): Result<PartBRecordId, List<Pair<DocumentCategory, FileName>>> =
+    recallRepository.getByRecallId(recallId).let { recall ->
+      // TODO: Move validation to DTO?  Even then, cross-field validation requires custom code.
+      //  As these properties are mutually dependent an alternative would be to combine them as a single object in the DTO
+      val oasysName = request.oasysFileName
+      val oasysContent = request.oasysFileContent
+      if ((oasysName == null && !oasysContent.isNullOrBlank()) || (oasysContent.isNullOrBlank() && oasysName != null)) {
+        throw IllegalArgumentException("Both fileName and fileContent of OASYS must be supplied as non-blank or neither")
+      }
+      val docs = scanAndStoreDocuments(request, recallId, currentUserId)
+
+      val savedFiles = mutableMapOf<DocumentCategory, DocumentId>()
+      val failedFiles = mutableListOf<Pair<DocumentCategory, FileName>>()
+      docs.forEach { result ->
+        result.value.second
+          .map { savedFiles.put(result.key, result.value.second.valueOrNull()!!) }
+          .mapFailure { failedFiles.add(result.key to result.value.first) }
+      }
+      if (failedFiles.isNotEmpty()) {
+        return Failure(failedFiles)
+      }
+      val currentVersion = recall.partBRecords.maxByOrNull { it.version }?.version ?: 0
+      val record = partBRecordRepository.save(
+        PartBRecord(
+          ::PartBRecordId.random(),
+          recallId,
+          request.details,
+          request.partBReceivedDate,
+          savedFiles[PART_B_RISK_REPORT]!!,
+          savedFiles[PART_B_EMAIL_FROM_PROBATION]!!,
+          savedFiles[OASYS_RISK_ASSESSMENT],
+          currentVersion + 1,
+          currentUserId,
+          OffsetDateTime.now()
+        )
+      )
+      Success(record.id())
+    }
+
+  private fun scanAndStoreDocuments(request: PartBRequest, recallId: RecallId, currentUserId: UserId):
+    Map<DocumentCategory, Pair<FileName, Result<DocumentId, VirusScanResult>>> {
+    val res = mutableMapOf(
+      scanAndStore(PART_B_RISK_REPORT, request.details, request.partBFileContent, request.partBFileName, recallId, currentUserId),
+      scanAndStore(PART_B_EMAIL_FROM_PROBATION, request.details, request.emailFileContent, request.emailFileName, recallId, currentUserId)
+    )
+    request.oasysFileContent?.let {
+      val oasysPair = scanAndStore(OASYS_RISK_ASSESSMENT, "Uploaded alongside Part B", it, request.oasysFileName!!, recallId, currentUserId)
+      res.put(oasysPair.first, oasysPair.second)
+    }
+    return res
+  }
+
+  private fun scanAndStore(
+    category: DocumentCategory,
+    details: String,
+    fileContent: String,
+    fileName: FileName,
+    recallId: RecallId,
+    currentUserId: UserId
+  ): Pair<DocumentCategory, Pair<FileName, Result<DocumentId, VirusScanResult>>> =
```

Comment:
🤮 . Does it need the filename too? The Pair -> Pair feels nasty.

Reply:
Jon wants the filename and category returned in the event of error: easiest way to do that looks like keeping them together in the returned Failure 🤷‍♂️ .

Reply:
Ok, not sure why you'd need the filename

Reply:
https://mojdt.slack.com/archives/C0223AGGQU8/p1648026203846159?thread_ts=1647360199.283689&cid=C0223AGGQU8